fms/training/plugins.py

```diff
@@ -240,7 +254,31 @@ def step(
         model_name = (
             model.__class__.__name__.lower() if self.name is None else self.name
         )
-        model_dict = model.state_dict()
+
+        # For FSDP, consolidate checkpointable data to rank0.
+        # TODO: may also want to support distcp checkpoints which should be faster
+        # to save and load but are harder to use for inference.
+
+        is_fsdp = isinstance(model, FSDP)
+        # For HSDP, self.group is only set for the first shard group. We only
+        # need to save checkpoints for one shard group.
+        if is_fsdp and self.group is None:
+            return
+
+        if is_fsdp:
+            dict_type = StateDictType.FULL_STATE_DICT
+            cfg = FullStateDictConfig(offload_to_cpu=True, rank0_only=True)
+            with dist.fsdp.FullyShardedDataParallel.state_dict_type(
+                model, dict_type, cfg
+            ):
+                print0("Aggregating FSDP model statedict")
+                model_dict = model.state_dict()
+                print0("Aggregating optim statedict")
+                optim_dict = FSDP.optim_state_dict(model, optimizer, group=self.group)
+        else:
+            model_dict = model.state_dict()
+            optim_dict = optimizer.state_dict()
+
```

Comment:
Slightly confused here - the approach being taken for FSDP is to accumulate everything on rank0, and then save from there. Is this not also the approach for HSDP? How do we get away with saving each individual model_dict for HSDP but not FSDP?

Reply:
here `is_fsdp = isinstance(model, FSDP)` , it's not differentiating fully vs hybrid sharding. The `else` block here is just for non-fsdp models, where the StateDictConfig etc aren't applicable.